pkg/controller/seldondeployment/seldondeployment_controller.go

```diff
@@ -699,51 +403,68 @@ func createComponents(r *ReconcileSeldonDeployment, mlDep *machinelearningv1alph
 				deploy.Spec.Replicas = &p.Replicas
 			}
 
-			// Add service orchestrator to first deployment if needed
-			if j == 0 && !hasSeparateEnginePod {
-				engineContainer, err := createEngineContainer(mlDep, &p, engine_http_port, engine_grpc_port)
-				if err != nil {
-					return nil, err
-				}
-				deploy.Labels[machinelearningv1alpha2.Label_svc_orch] = "true"
-
-				deploy.Spec.Template.Spec.Containers = append(deploy.Spec.Template.Spec.Containers, *engineContainer)
-				//deploy.Spec.Template.Spec.ServiceAccountName = getEnv("ENGINE_CONTAINER_SERVICE_ACCOUNT_NAME", "seldon")
-				//deploy.Spec.Template.Spec.DeprecatedServiceAccount = deploy.Spec.Template.Spec.ServiceAccountName
-				//deploy.Spec.Template.Annotations = map[string]string{}
-				if deploy.Spec.Template.Annotations == nil {
-					deploy.Spec.Template.Annotations = make(map[string]string)
-				}
-				//overwrite annotations with predictor annotations
-				for _, ann := range p.Annotations {
-					deploy.Spec.Template.Annotations[ann] = p.Annotations[ann]
+			// create services for each container
+			for k := 0; k < len(cSpec.Spec.Containers); k++ {
+				var con *corev1.Container
+				con = utils.GetContainerForDeployment(deploy, cSpec.Spec.Containers[k].Name)
+
+				// engine will later get a special predictor service as it is entrypoint for graph
+				// and no need to expose tfserving container as it's accessed via proxy
+				if con.Name != "seldon-container-engine" && con.Name != "tfserving" && con.Name != "" {
+
+					// service for hitting a model directly, not via engine
+					svc := createContainerService(deploy, p, mlDep, con, c)
+					if svc != nil {
+						c.services = append(c.services, svc)
+					} else {
+						// a user-supplied container may not be a pu so we may not create service for that
+						log.Info("Not creating container service for " + con.Name)
+					}
 				}
-				// Add prometheus annotations
-				deploy.Spec.Template.Annotations["prometheus.io/path"] = getEnv("ENGINE_PROMETHEUS_PATH", "/prometheus")
-				deploy.Spec.Template.Annotations["prometheus.io/port"] = strconv.Itoa(engine_http_port)
-				deploy.Spec.Template.Annotations["prometheus.io/scrape"] = "true"
-
-				deploy.ObjectMeta.Labels[machinelearningv1alpha2.Label_seldon_app] = pSvcName
-				deploy.Spec.Selector.MatchLabels[machinelearningv1alpha2.Label_seldon_app] = pSvcName
-				deploy.Spec.Template.ObjectMeta.Labels[machinelearningv1alpha2.Label_seldon_app] = pSvcName
-
 			}
+			if len(deploy.Spec.Template.Spec.Containers) > 0 && deploy.Spec.Template.Spec.Containers[0].Name != "" {
```

Comment:
Can only happen when there's no container spec because the deployment to be created is using prepackaged. So then it goes through this loop but doesn't create anything in the deployment. Am adding a comment to clarify.

Reply:
If we've ended up with an empty deployment it means it's a prepackaged use-case and will be handled later in the code. Am adding comment.